code/parse/sexp.cpp

```diff
@@ -10455,14 +10455,21 @@ int eval_for_counter(int arg_handler_node, int condition_node, bool just_count =
 	// build a vector of counter values
 	for (i = counter_start; ((counter_step > 0) ? i <= counter_stop : i >= counter_stop); i += counter_step)
 	{
-		sprintf(buf, "%d", i);
-		argument_vector.emplace_back(vm_strdup(buf), -1);
-		// Note: we do not call vm_free() on the contents of argument_vector, and we don't for the very good
-		// reason that those pointers are then passed along by test_argument_vector_for_condition(), below.
-		// The strings will then be freed by arg_item::expunge() or arg_item::clear_nesting_level(), or even
-		// inside test_argument_vector_for_condition() (if the argument doesn't satisfy the condition). So
-		// under no circumstances try to free these strings inside this function! It will cause a double-free
-		// situation, resulting in a crash at best. -MageKing17
+		// if we're just counting we should just avoid the memory management problem mentioned below,
+		// since we return before calling test_argument_vector_for_condition
+		if (just_count)
+			argument_vector.emplace_back("", -1);
```

Comment:
Won't this call result in dangling pointers?

Where are the empty strings we're adding here being stored?

Reply:
Strings that are hard-coded do not end up as dangling pointers.  They are presumably kept in persistent area of the code.